CISSSO Integration (Quality of Life Updates to analysis.py)

petruong opened the pull request:
Hello there. Here at CISSSO, we've been using/updating from a version of OMG Film Dosimetry python modules prior to its introduction here onto Github. Going to try to keep the summarized changes brief and focus on the major updates/feature revisions.
Additional information (i.e., thought process/explanation for re-work) can be found in commit descriptions.  

General formatting:
- New line statements to improve print readability
- Tried to maintain function/script description/commentary

analysis.py:
- norm_film_dose integration: added the possibility to use a separate scan for the normalization film (i.e., 300 MU). If exists, it will be used to acquire the reference normalization factor. Having a separate scan can be useful if multiple patient QA has been scanned during the same day, to which the separate normalization scan (scan image files) can be re-used.  Another use would be if there exists lateral/longitudinal scanner response issues where scanning in the center of the scanner (for both the QA and normalization film) would decrease this response effect.
- Fixed up some issues with closing the self.fig (commonly found in functions depending on an event/action) prior to the call for plt.wait, where there will be a missing fig/plt instance pointing to it. Spyder is able to work around this, but other IDE unfortunately will crash as a consequence. 
  - Updated the plt.wait statement to be 1 second as opposed to 5 second. Quicken the workflow a bit (definitely more convenient when testing code, but also nice to just be generally faster). 
- Vertical line present in profiles created (during show_results) to give a better idea of its relation to the crosshair (profile lines) drawn on the results (film, reference, gamma, RSE plots)
- Inclusion of filename of the film and reference dose in the results figure title per axis
   - Inclusion of the particular gamma analysis criteria in the title of the results figure
- select_markers function has been reworked to simplify view and general usage
   - Limit number of marker selection to 4
   - Fixed a bug where ax.clear() would occur when pressing "hotkeys" for zoom/full-screen (and obviously other tools). This would conflict with the self.film_dose.array manipulations (rotate, flip, etc.). 
      - Removed markers when these manipulations occur as the array/figure has changed
   - Added an option where if no markers are selected, the top, right, bottom, left (center) sides have been selected
      - Not perfect, but at least it serves as an option where markers were forgotten/omitted during the irradiation step. Further registration position tuning can be done to further align.
- set_profiles function has been reworked that now includes clicking in the profile to change the view in the other profile (vertical line: ultimately moving the crosshair as well)
   - Required introducing a self.prof_x, y variable to store the profile position
   - Cleaned up set_profiles function in general to simply be based on the cursor type and a click

calibration.py
- Added default parameter for variable savefile in show_results function call. When working with the calibration python module (command line prompts), not relevant to have to the savefile save/exist when wanting to view the results.

tiff2dose.py
- Updated the clim range for consistency, disturbance, residuals as the previous values did not show much relevant information that could be extracted

OMG_CISSSO_Test.py:
- Included example QA patient tiff2analysis test file to demonstrate how the OMG film dosimetry python module is implemented/utilized at CISSSO.
- Useful implementation would be the selection of the scan, reference dose file path
- Future implementation would be to fully integrate the input for info variable, LUT selection, other tiff2dose parameters perhaps.

Happy reviewing.
Cheers,
Peter

**Edit:** didn't realize that by pushing new commits from my local repository would cause it to be included in the pull request... Whoops. Anyways, a summary (particular changes are described in the commit notes):
Analysis_Pickle is designed to simply load in .pkl files (calibration, tiff2dose, or tiff2analysis/dose2analysis) and simply show the results. It has been mainly been used as a debugging tool (since it will reload the src scripts when loading the pkl object: calibration, tiff2dose, analysis). It can serve as a standalone tool (in the form of a batch file to launch outside of the Spyder environment, but within Windows calling Python itself) to "load" in resultant tiff2analysis/dose2analysis pickled files (if you're including this in your workflow) to simply illustrate the results (yet allowing interactive features such as viewing the dose, gamma, profiles, etc. as opposed to a static result published by the pdf).

**Edit 2:** per request by JF, I have re-converted the profile pixel position back to mm. Updated what happens to xdata when a click on the profile is done to re-convert the mm x_axis back into pixel position. More details can be found in the commit description (2023-12-15).

petruong committed:
Update calibration.py

calibration.py
- Added (commented out for reference) lines for changes needed for the profile adjustments (Eclipse/CISSSO). Ended up not using these changes anyways as lateral correction was applied for our center.
- Added default parameter for variable savefile in show_results function call. Useful when debugging calibration LUT object to view results without necessarily needing to save the figure to a file.

OMG_CISSSO_Test.py
- QA Patient analysis file used at CISSSO. Will be using it to compare the changes needed to implement into the Github v1.5.1 version of OMG_Dosimetry.

petruong committed:
analysis.py (CISSSO Integration)

- norm_film_dose (variable used in QApatient script on if the normalization film [i.e., 300 MU] was scanned separately than the patient QA scan). included in DoseAnalysis initialization. Idea with scanning normalization film separately is if the user would like to have the patient QA film scanned more in the center of the scanner (along with the normalization film) if there are issues with lateral response issues associated with the scanner.
 - Added lines of norm_film_dose should not affect rest of script functionality as by default, its initialization value is None.

- Modified/added plt.close(self.fig) lines following while self.wait calls to close figure instance/pointers. Found that there were issues in IDEs other than Spyder where the figure remained opened/script would not proceed because a pointer was trying to access a different figure.
- Commented out a few instances of plt.close(self.fig) lines that were in key_press event-like functions where self.wait is changed to True, but doesn't make sense to close these figures before the while self.wait is done. Review comment above indicating the added plt.close(self.fig) following directly after the while loop is complete. Issue was that the figure would try to close, but it was already closed before in the key_press event function.
- Moved self.tune_registration() function call to the end of the register() rather than be at the end of the event type call. Makes more sense rather than be reliant on the end of an event.

- Included a vertical line that can be visualized on the profile plots indicating where the cursor is
 - Rather than have the x_axis plot based on a conversion to mm, have it based on pixel position as the rest of the plots are in pixel position.
- Included the name of the film and reference dose files in the title of the plots.

- Added a few comments in the move_iso_center function to be reviewed (since I am currently not using the markers_center variable as it wasn't quite clear the translation from Eclipse vs. Raystation, I have left the functionality of move_iso_center the same). Depends on if it's the same post-export from Raystation, the orientation of the reference dose DICOM could be extracted from the ImageOrientationPatient tag based on the corner matrix rather than a series description (which is not output by default in Eclipse)

petruong committed:
analysis.py (figure window title)

- Depreciated: fig.canvas.set_window_title:
 - Updated to fig.canvas.manager.set_window_title

- Tested with new changes. Everything seems to be functioning other than the bug with the select_markers function when moving the cursor outside of the selection FOV for the markers.

petruong committed:
tiff2dose.py (clim changes: consistency/delta/residual)

- Default clim settings for Consistency, Disturbance, Residuals do not show much information.
 - Without having to manually change the clim values, I have set the clim values based on the values present specific to the relevant dose consistency, delta, residual error arrays.

petruong committed:
analysis (select_markers axis clear bug)

- Assigned cursor variable/object to plt rather than have an unused variable (similar to how it is done in show_profiles function for the plt.multi object (line 755).

- Moved the clearing of the fig when a key has been pressed (i.e., "o" for zoom, "f" for full-screen to the relevant keypresses (up, left, right) that would need the clearing/overwrite of the fig.
- Added a threshold in how many markers that can be selected.
- Simplified a bit plotting of markers

- Change pause time to 1 to be faster when moving to the next step (removes lag)
 - Added corresponding return statements to follow same style to other end of self.wait loops

petruong committed:
analysis (set markers/profile rework)

- General clean-up of comments/old CISSSO code (already have a copy/saved in previous commits)
 - matplotlib is already utilized in other calls (removed due to redundacy)
 - Following single-line if/else conditional statements

- set_profiles:
 - Removed zooming_panning variable and simply have the changing of profiles based on if mouse was clicked and if cursor was indeed an arrow.
 - Introduced self.prof_x and self.prof_y  as profile x, y positions that stores and saves it for next clicks/profile selection
 - Clicking of the profile will now move the other profile based on the X-axis
 - Added a message indicating that zoom/pan tool is active and cannot select profile as a result.

- select_markers/ontype:
 - Re-organized the conditional for event.key == "enter".
 - Not perfect, but if no markers were selected, take the center of the self.film_dose.array (needs to be adjusted in the tune_registration step, but at least it is possible if no markers were on the film or selected by user)

petruong committed:
Update analysis.py

Quick clean-up

petruong committed:
CISSSO (Pull Request Cleanup)

OMG_CISSSO_Test.py
- Including the changes made to this test  QA script just to have an example of how src scripts have been utilized at the CISSSO stage/level

analysis/calibration/tiff2dose.py
- Updated header structure for uniform versioning status/date
- Removed CISSSO specific comments (as I already have these notes in my own scripts/commits) with the aim of not having too much reviewing during the pull request stage.

petruong committed:
Post Compare (Cleanup)

analysis.py
- Updated commentary/description for functions/methods so that they retain same structure
- Removed local variable film/ref_fileName as this can be called directly when adding the title
- Updated plot_markers function to not plot each marker over and over within the for loop, but rather do the last marker (len(self.markers)).
- Added new line statements in relevant print calls (To separate lines in the print window to improve readability/messages intended for the user to read to understand the next step)

OMG_CISSSO_Test.py
- Changed initial directory to go to directory above after selecting scan (for better flow when selecting the DICOM reference dose file)

petruong committed:
Update analysis.py

Typo \m versus \n before pull request, whoops.

LuisOlivaresJ commented:
Hi there,

I only have some experience using calibration.py. The changes there are ok to me.

petruong committed:
Analysis_Pickle Implementation

Analysis_Pickle is designed to simply load in .pkl files (calibration, tiff2dose, or tiff2analysis/dose2analysis) and simply show the results. It has been mainly been used as a debugging tool (since it will reload the src scripts when loading the pkl object: calibration, tiff2dose, analysis). It can serve as a standalone tool (in the form of a batch file to launch outside of the Spyder environment, but within Windows calling Python itself) to "load" in resultant tiff2analysis/dose2analysis pickled files (if you're including this in your workflow) to simply illustrate the results (yet allowing interactive features such as viewing the dose, gamma, profiles, etc. as opposed to a static result published by the pdf).

Changes made to src scripts to correct for the "bugs" allowed by Spyder are as follows:

imageRGB.py
- plotCB function uses plt.show() which pauses the script (issue when calling plotCB(...) in show_results() functions) until the figure has been closed. Using the parameter block = False removes the issue of the figure blocking the script from moving forward whislt still plotting.

analysis.py
- An issue with show_profiles function calling plt.show() prior to canvas connect button_press_event.
- Added a default parameter for show = True for this plt.show() call to not disrupt its calling in other scripts
 - Changed its usage in publish_pdf to not show the figure as its main goal is to publish the pdf.

calibration.py
- Included the show parameter similar design as analysis.py
 - Grouped up related/repeated code and placed outside of if/else condition
- Similar concept for show = False in publish_pdf function

tiff2dose.py
- Similar thought process for show results show = True default parameter
 - Included multi cursor as a bonus, but not sure of its relevance
- publish_pdf to have the same show = False as no need to show figure when endgoal is to create the pdf

Analysis_Pickle.bat
- Command line parameters need to be changed in order to be used
- First parameter is the location of the python compiler executable
- -i is to be used in interactive mode
- second/last parameter is the location of the Analyis_Pickle.py script
- Note: absolute paths used are in ANSI format due to French accents.

petruong commented:
Note: pull request has been updated from my recent commit (2023-12-06)... 
Review edited description (bottom) for additional details.

jfcabana commented on src/omg_dosimetry/analysis.py:
```diff
@@ -608,166 +623,175 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
                 position = np.floor(self.ref_dose.shape[0] / 2).astype(int)
-            film_prof = film[position,:]
-            ref_prof = ref[position,:]
-            x_axis = (np.array(range(0, len(film_prof))) / self.film_dose.dpmm).tolist()
```
I think it would be better to leave the profiles x_axis to mm. It used to be in pixels, but I changed it to mm because we want to be able use this axis to estimate a physical offset between the film dose profile and reference dose profile. 

If it is bothering to have different axis for the dose plots and profile plots, it would make more sense to change the other  plots to mm.

After testing, I realize that changing to pixels also screws up my profile offsets analysis, as the move_profile_ontype function would now apply shifts by 0.1 pixel increment instead of 0.1 mm as intended.

petruong replied:
Updated from recent commit (2023-12-15). 
Also to note: the differing axis (pixel #/position) for the dose and profile plots is a hard one to tackle since it is generated with imageRGB.plotCB() using matplotlib.pyplot.imshow() for the 2D distribution, the axes would need to be carefully looked at during the conversion. 
Since it is not that much of an issue, I will not look at it for this pull request and to simply focus on conversion back to mm from pixel #/position.

jfcabana commented on src/omg_dosimetry/analysis.py:
```diff
@@ -608,166 +623,175 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
+                if profile == 'x': title='Profile horizontal (y={})'.format(position)
+                if profile == 'y': title='Profile vertical (x={})'.format(position)
         ax.set_title(title)
-        ax.set_xlabel('Position (mm)')
```
:(

jfcabana reviewed (changes requested):
Seems all good, except for the changing of x_axis in profiles plots to pixels. See comments below.

petruong committed:
analysis.py (plot_profile: pixel to mm)

- Organized/cleaned up plot_profile function by following similar multi-initialization declarations on single line.
- Removed repeatable if/else conditional lines based on if profile was a single int or an array of int values.
- Converted pixel position values into mm (as requested).
 - As a consequence, updated the click x.data to be reconverted back into pixel position when clicking within the profile axes (axes[4, 5])
- Updated/standardized capitalization for figure titles

*** Standardized the parameter name in plot_gamma_stats using show_var_DoseTA to show_varDoseTA to match the already used show_varDistTA. This may break some patient QA scripts that are using show_var_DoseTA as a parameter. An example of it being used/changed is found in the OMG_CISSSO.Test.py script

jfcabana reviewed (approved):
All good!
